### PR TITLE
feat(lua): vim.ui_attach to get ui events from lua

### PR DESCRIPTION
**Pull request by bfredl**
somewhat inspired by #16383 but I was thinking of an alternative which (1) is generic over all ext_ "widgets" (i e cmdline, popupmenu and messages) and (2) if vim compat is not a direct goal then we could use something lua based.

As an example this gets popupmenu events only, which perhaps can simplify plugins that implement custom popupmenu as float:
```
vim.ui_attach({ext_popupmenu=true}, function (a, ...)
  require'luadev'.print(a, vim.inspect{...})
end)
```

### CODE REVIEW

File: src/nvim/ui_compositor.c [hunk 1]
```diff
@@ -666,3 +672,20 @@ static void ui_comp_grid_resize(UI *ui, Integer grid, Integer width, Integer hei
   }
 }
 
+static void ui_comp_event(UI *ui, char *name, Array args, bool *args_consumed)
```
Comment (bfredl) [inline src/nvim/ui_compositor.c:675]: note: using the compositor for this is somewhat a hack but it enables quick prototyping and you could argue it is used primarily for composed ui:s (i e TUI) as multigrid ui:s will intend to implement external widgets themselves.

**Comment by vigoux** [comment]
When trying this PR with `ext_messages`, it seems that:
- The cursor is moved to the now-gone command-line when entering any command
- I don't get any update during the time of the command

**Comment by bfredl** [comment]
currently ext_messages has to imply ext_cmdline as the cmdline uses message functions internally. I've pushed a fix that should remove the incorrect cursor behaviour. Although in my upcoming `cmdheight=0` rework, I will see if we can lift this limitation initially.

**Comment by folke** [comment]
> currently ext_messages has to imply ext_cmdline as the cmdline uses message functions internally. I've pushed a fix that should remove the incorrect cursor behaviour. Although in my upcoming `cmdheight=0` rework, I will see if we can lift this limitation initially.

Ideally, the cursor should be hidden when implementing `ext_cmdline` :)
See #20309

**Pull request merged**